app/api/v1/auth/index.js

```diff
@@ -82,16 +83,51 @@ router.post('/register', (req, res, next) => {
     return next(new error.BadRequest('Password must be at least 10 characters long'));
   }
 
-  // TODO account activation via email
+
   const newUser = User.sanitize(req.body.user);
   newUser.state = 'ACTIVE';
   User.generateHash(req.body.user.password).then((hash) => {
     newUser.hash = hash;
-    return User.create(newUser);
-  }).then((user) => {
-    log.info('user authentication (registration)', { request_id: req.id, user_uuid: user.uuid });
-    res.json({ error: null, user: user.getPublicProfile() });
-    Activity.accountCreated(user.uuid);
+    // require that we create a user succesfully and send email succesfully
+    db.transaction({
+      isolationLevel: Sequelize.Transaction.ISOLATION_LEVELS.REPEATABLE_READ,
+    }, (transaction) => Promise.all([
+      User.create(newUser).then((user) => {
+        User.generateAccessCode().then((code) => {
+          user.accessCode = code;
+          email.sendEmailVerification(user.email, user.firstName, code).then(() => {
+            // save the access code if email was sent
+            user.save();
```

Comment:
This writes to the same user twice, and I get the rationale behind saving the access code once the email's sent, but it doesn't change anything functionally if we do this double write and or if we create the user including the access code (i.e. User::generateAccessCode &rarr; User::create &rarr; Email::sendEmailVerification) besides internally writing an access code that gets overwritten once the user re-requests email verification. Also the `save` returns a Promise, so if that fails, the transaction rolls back the entire user object but still sends an email, so if we want to only commit the transaction when the email is sent, `sendEmailVerification` should be the last Promise in the chain.

Although the `save` isn't currently being returned so if it failed the transaction would still go through, minus the access code, and the email would still be sent, so the email itself would be invalid since there's no access code written to the user object.

Marginal, basically negligible difference so I think we might as well simplify where we can and do a single database operation.

Reply:
did single